import_obfuscation.go

```diff
@@ -0,0 +1,572 @@
+	// skip test symbols
+	"_test.",
+}
+
+// prefixes of symbols that we will garble,
+// but we split the symbol name by one of
+// these prefixes so that we do not
+// accidentally garble an essential prefix
+var symPrefixes = [...]string{
+	"go.builtin.",
+	"go.itab.",
+	"go.itablink.",
+	"go.interface.",
+	"go.map.",
+	"gofile..",
+	"type..eq.",
+	"type..eqfunc.",
+	"type..hash.",
+	"type..importpath.",
+	"type..namedata.",
+	"type.",
+}
+
+// splitSymbolPrefix returns the symbol name prefix Go uses
+// to help designate the type of the symbol, and the rest of
+// the symbol name. Additionally, a bool is returned that
+// signifies whether garbling the symbol name should be skipped.
+func splitSymbolPrefix(symName string) (string, string, bool) {
+	if symName == "" {
+		return "", "", true
+	}
+
+	for _, prefix := range skipPrefixes {
+		if strings.HasPrefix(symName, prefix) {
+			return "", "", true
+		}
+	}
+
+	for _, entrySym := range entrypointSyms {
+		if symName == entrySym {
+			return "", "", true
+		}
+	}
+
+	for _, substr := range skipSubstrs {
+		if strings.Contains(symName, substr) {
+			return "", "", true
+		}
+	}
+
+	for _, prefix := range symPrefixes {
+		if strings.HasPrefix(symName, prefix) {
+			return symName[:len(prefix)], symName[len(prefix):], false
+		}
+	}
+
+	return "", symName, false
+}
+
+// privateImportIndex returns the offset and length of a private import
+// in symName. If no private imports from privImports are present in
+// symName, -1, 0 is returned.
+func privateImportIndex(symName string, privImports privateImports, nameDataSym bool) (int, int) {
+	matchPkg := func(pkg string) int {
+		off := strings.Index(symName, pkg)
+		if off == -1 {
+			return -1
+			// check that we didn't match inside an import path. If the
+			// byte before the start of the match is not a small set of
+			// symbols that can make up a symbol name, we must have matched
+			// inside of an ident name as a substring. Or, if the byte
+			// before the start of the match is a forward slash, we are
+			// definitely inside of an input path.
+		} else if off != 0 && (!isSymbol(symName[off-1]) || symName[off-1] == '/') {
+			return -1
+		}
+
+		return off
+	}
+
+	firstOff, l := -1, 0
+	for _, privatePkg := range privImports.privatePaths {
+		off := matchPkg(privatePkg)
+		if off == -1 {
+			continue
+		} else if off < firstOff || firstOff == -1 {
+			firstOff = off
+			l = len(privatePkg)
+		}
+	}
+
+	if nameDataSym {
+		for _, privateName := range privImports.privateNames {
+			// search for the package name plus a period, to
+			// minimize the likelihood that the package isn't
+			// matched as a substring of another ident name.
+			// ex: pkgName = main, symName = "domainname"
+			off := matchPkg(privateName + ".")
+			if off == -1 {
+				continue
+			} else if off < firstOff || firstOff == -1 {
+				firstOff = off
+				l = len(privateName)
+			}
+		}
+	}
+
+	return firstOff, l
+}
+
+func isSymbol(c byte) bool {
```

Comment:
add a comment

Reply:
The logic for this func is commented at line 443, the only place it's called